[Bug]: Severe Memory Leak in core utilities via @lru_cache on instance methods

### Version

0.1.dev

### Which installation method(s) does this occur on?

Source

### Describe the bug.

In Python, using `functools.lru_cache` on a class instance method creates a permanent, strong reference to `self` in the cache keys of the class-level function object. Because the cache persists for the lifetime of the program, the class instances can **never** be garbage collected.

Running static analysis (`ruff check --select B`) on the `NeMo-Agent-Toolkit` codebase reveals **25 occurrences** of this `B019` anti-pattern across critical core utilities:
1. `FunctionBase` in `function_base.py` (9 occurrences)
2. `DecomposedType` in `type_utils.py` (11 occurrences)
3. `TypeIntrospectionMixin` in `type_introspection_mixin.py` (5 occurrences)

Because `FunctionBase` and `DecomposedType` wrapper instances are created constantly during type inspection (which is used heavily in validation, routing, and telemetry during server operation), these leaked instances accumulate permanently in long-running processes like `nat mcp serve` or `nat serve`. By default, `lru_cache` locks the 128 most recent instances in memory per decorated property (i.e. `25 * 128 = 3200` permanent objects locked in memory).

**Expected Behavior:**
Core utility instances should be garbage collected once they go out of scope. Property wrappers should use `functools.cached_property` (which correctly stores the cache in the instance `__dict__` so it is garbage collected along with the instance), or remove the cache if the operation is cheap.

### Minimum reproducible example

```shell
import gc
from pydantic import BaseModel
from nat.builder.function_base import FunctionBase

class DummyFunction(FunctionBase[str, str, str]):
    def __init__(self):
        super().__init__()

# Create 10,000 instances and access a cached property
for _ in range(10000):
    obj = DummyFunction()
    _ = obj.input_type
    
del obj

# Force garbage collection
gc.collect()

# Count how many objects are permanently locked in memory
alive_objects = sum(1 for o in gc.get_objects() if isinstance(o, DummyFunction))
print(f"Leaked instances: {alive_objects}") # Will print exactly 128
```

### Relevant log output

<details><summary>Click here to see error details</summary><pre>

uv run ruff check --select B019 packages/nvidia_nat_core/src/nat

packages/nvidia_nat_core/src/nat/builder/function_base.py:94:14: B019 Use of `functools.lru_cache` or `functools.cache` on methods can lead to memory leaks
packages/nvidia_nat_core/src/nat/builder/function_base.py:121:14: B019 Use of `functools.lru_cache` or `functools.cache` on methods can lead to memory leaks
packages/nvidia_nat_core/src/nat/builder/function_base.py:146:14: B019 Use of `functools.lru_cache` or `functools.cache` on methods can lead to memory leaks
... (25 total B019 warnings)

</pre></details>

### Other/Misc.

**Proposed Solution:**
I can raise a PR to safely refactor these occurrences:
1. Replace all `@property` + `@lru_cache` combinations with `@functools.cached_property`.
2. For any remaining `@lru_cache` instances on plain methods (e.g. `_extract_input_output_types` or args-taking methods like `_get_union_info`), switch to using an internal instance dictionary cache (e.g., `self._cache[arg]`) or drop the cache entirely if the operation is cheap enough to rerun safely.

### Code of Conduct

- [x] I agree to follow the NeMo Agent Toolkit Code of Conduct
- [x] I have searched the [open bugs](https://github.com/NVIDIA/NeMo-Agent-Toolkit/issues?q=is%3Aopen+is%3Aissue+label%3Abug) and have found no duplicates for this bug report

--- a/packages/nvidia_nat_core/src/nat/builder/function_base.py
+++ b/packages/nvidia_nat_core/src/nat/builder/function_base.py
@@ -23,13 +23,13 @@ import logging
 import typing
 from abc import ABC
 from collections.abc import Callable
-from functools import lru_cache
 from types import NoneType
 
 from pydantic import BaseModel
 
 from nat.utils.type_converter import TypeConverter
 from nat.utils.type_utils import DecomposedType
+from nat.utils.type_utils import read_only_cached_property
 
 InputT = typing.TypeVar("InputT")
 StreamingOutputT = typing.TypeVar("StreamingOutputT")
@@ -92,8 +92,7 @@ class FunctionBase(typing.Generic[InputT, StreamingOutputT, SingleOutputT], ABC)
 
         self._converter: TypeConverter = TypeConverter(converters)
 
-    @property
-    @lru_cache
+    @read_only_cached_property
     def input_type(self) -> type[InputT]:
         """
         Get the input type of the function. The input type is determined by the generic parameters of the class.
@@ -120,8 +119,7 @@ class FunctionBase(typing.Generic[InputT, StreamingOutputT, SingleOutputT], ABC)
 
         raise ValueError("Could not find input schema")
 
-    @property
-    @lru_cache
+    @read_only_cached_property
     def input_class(self) -> type:
         """
         Get the python class of the input type. This is the class that can be used to check if a value is an instance of
@@ -146,8 +144,7 @@ class FunctionBase(typing.Generic[InputT, StreamingOutputT, SingleOutputT], ABC)
 
         return input_origin
 
-    @property
-    @lru_cache
+    @read_only_cached_property
     def input_schema(self) -> type[BaseModel]:
         """
         Get the Pydantic model schema for validating inputs. The schema must be pydantic models. This allows for
@@ -179,8 +176,7 @@ class FunctionBase(typing.Generic[InputT, StreamingOutputT, SingleOutputT], ABC)
         """
         return self._converter_list
 
-    @property
-    @lru_cache
+    @read_only_cached_property
     def streaming_output_type(self) -> type[StreamingOutputT]:
         """
         Get the streaming output type of the function. The streaming output type is determined by the generic parameters
@@ -208,8 +204,7 @@ class FunctionBase(typing.Generic[InputT, StreamingOutputT, SingleOutputT], ABC)
 
         raise ValueError("Could not find output schema")
 
-    @property
-    @lru_cache
+    @read_only_cached_property
     def streaming_output_class(self) -> type:
         """
         Get the python class of the output type. This is the class that can be used to check if a value is an instance
@@ -234,8 +229,7 @@ class FunctionBase(typing.Generic[InputT, StreamingOutputT, SingleOutputT], ABC)
 
         return output_origin
 
-    @property
-    @lru_cache
+    @read_only_cached_property
     def streaming_output_schema(self) -> type[BaseModel] | type[None]:
         """
         Get the Pydantic model schema for validating streaming outputs. The schema must be pydantic models. This allows
@@ -255,8 +249,7 @@ class FunctionBase(typing.Generic[InputT, StreamingOutputT, SingleOutputT], ABC)
         """
         return self._streaming_output_schema
 
-    @property
-    @lru_cache
+    @read_only_cached_property
     def single_output_type(self) -> type[SingleOutputT]:
         """
         Get the single output type of the function. The single output type is determined by the generic parameters
@@ -284,8 +277,7 @@ class FunctionBase(typing.Generic[InputT, StreamingOutputT, SingleOutputT], ABC)
 
         raise ValueError("Could not find output schema")
 
-    @property
-    @lru_cache
+    @read_only_cached_property
     def single_output_class(self) -> type:
         """
         Get the python class of the output type. This is the class that can be used to check if a value is an instance
@@ -310,8 +302,7 @@ class FunctionBase(typing.Generic[InputT, StreamingOutputT, SingleOutputT], ABC)
 
         return output_origin
 
-    @property
-    @lru_cache
+    @read_only_cached_property
     def single_output_schema(self) -> type[BaseModel] | type[None]:
         """
         Get the Pydantic model schema for validating single outputs. The schema must be pydantic models. This allows for

--- a/packages/nvidia_nat_core/src/nat/observability/mixin/type_introspection_mixin.py
+++ b/packages/nvidia_nat_core/src/nat/observability/mixin/type_introspection_mixin.py
@@ -16,7 +16,7 @@
 import inspect
 import logging
 import types
-from functools import lru_cache
+from collections.abc import Callable
 from typing import Any
 from typing import TypeVar
 from typing import get_args
@@ -38,6 +38,29 @@ class TypeIntrospectionMixin:
     This mixin combines the DecomposedType class utilities with MRO traversal
     to properly handle complex inheritance chains like HeaderRedactionProcessor or ProcessingExporter.
     """
+
+    def _get_or_set_cache(self, attr_name: str, compute_func: Callable[[], Any], key: Any = None) -> Any:
+        """Centralized helper for lazy cache evaluation.
+
+        Args:
+            attr_name (str): The name of the instance attribute to use for caching.
+            compute_func (Callable[[], Any]): Function to compute the value if not cached.
+            key (Any, optional): If provided, the cache is treated as a dictionary. Defaults to None.
+
+        Returns:
+            Any: The cached or newly computed value.
+        """
+        if key is None:
+            if not hasattr(self, attr_name):
+                setattr(self, attr_name, compute_func())
+            return getattr(self, attr_name)
+
+        if not hasattr(self, attr_name):
+            setattr(self, attr_name, {})
+        cache_dict = getattr(self, attr_name)
+        if key not in cache_dict:
+            cache_dict[key] = compute_func()
+        return cache_dict[key]
 
     def _extract_types_from_signature_method(self) -> tuple[type[Any], type[Any]] | None:
         """Extract input/output types from the signature method.
@@ -277,7 +300,6 @@ class TypeIntrospectionMixin:
 
         return None
 
-    @lru_cache
     def _extract_input_output_types(self) -> tuple[type[Any], type[Any]]:
         """Extract both input and output types using available approaches.
 
@@ -287,19 +309,19 @@ class TypeIntrospectionMixin:
         Raises:
             ValueError: If types cannot be extracted
         """
-        # First try the signature-based approach
-        result = self._extract_types_from_signature_method()
-        if result:
-            return result
 
-        # Fallback to MRO-based approach for complex inheritance
-        result = self._extract_instance_types_from_mro()
-        if result:
-            return result
+        def compute():
+            result = self._extract_types_from_signature_method()
+            if result:
+                return result
+            result = self._extract_instance_types_from_mro()
+            if result:
+                return result
+            raise ValueError(f"Could not extract input/output types from {self.__class__.__name__}. "
+                             f"Ensure class inherits from a generic like Processor[InputT, OutputT] "
+                             f"or has a signature method with type annotations")
 
-        raise ValueError(f"Could not extract input/output types from {self.__class__.__name__}. "
-                         f"Ensure class inherits from a generic like Processor[InputT, OutputT] "
-                         f"or has a signature method with type annotations")
+        return self._get_or_set_cache("_extract_input_output_types_cache", compute)
 
     @property
     def input_type(self) -> type[Any]:
@@ -319,7 +341,6 @@ class TypeIntrospectionMixin:
         """
         return self._extract_input_output_types()[1]
 
-    @lru_cache
     def _get_union_info(self, type_obj: type[Any]) -> tuple[bool, tuple[type, ...] | None]:
         """Get union information for a type.
 
@@ -329,8 +350,12 @@ class TypeIntrospectionMixin:
         Returns:
             tuple[bool, tuple[type, ...] | None]: (is_union, union_types_or_none)
         """
-        decomposed = DecomposedType(type_obj)
-        return decomposed.is_union, decomposed.args if decomposed.is_union else None
+
+        def compute():
+            decomposed = DecomposedType(type_obj)
+            return (decomposed.is_union, decomposed.args if decomposed.is_union else None)
+
+        return self._get_or_set_cache("_get_union_info_cache", compute, key=type_obj)
 
     @property
     def has_union_input(self) -> bool:
@@ -423,25 +448,31 @@ class TypeIntrospectionMixin:
 
         return False
 
-    @lru_cache
     def _get_input_validator(self) -> type[BaseModel]:
         """Create a Pydantic model for validating input types.
 
         Returns:
             type[BaseModel]: The Pydantic model for validating input types
         """
-        input_type = self.input_type
-        return create_model(f"{self.__class__.__name__}InputValidator", input=(input_type, FieldInfo()))
 
-    @lru_cache
+        def compute():
+            input_type = self.input_type
+            return create_model(f"{self.__class__.__name__}InputValidator", input=(input_type, FieldInfo()))
+
+        return self._get_or_set_cache("_input_validator_cache", compute)
+
     def _get_output_validator(self) -> type[BaseModel]:
         """Create a Pydantic model for validating output types.
 
         Returns:
             type[BaseModel]: The Pydantic model for validating output types
         """
-        output_type = self.output_type
-        return create_model(f"{self.__class__.__name__}OutputValidator", output=(output_type, FieldInfo()))
+
+        def compute():
+            output_type = self.output_type
+            return create_model(f"{self.__class__.__name__}OutputValidator", output=(output_type, FieldInfo()))
+
+        return self._get_or_set_cache("_output_validator_cache", compute)
 
     def validate_input_type(self, item: Any) -> bool:
         """Validate that an item matches the expected input type using Pydantic.
@@ -477,7 +508,6 @@ class TypeIntrospectionMixin:
             logger.warning("Item %s is not compatible with output type %s", item, self.output_type)
             return False
 
-    @lru_cache
     def extract_non_optional_type(self, type_obj: type | types.UnionType) -> Any:
         """Extract the non-None type from Optional[T] or Union[T, None] types.
 
@@ -490,7 +520,11 @@ class TypeIntrospectionMixin:
         Returns:
             Any: The actual type without None, or the original type if not a union with None
         """
-        decomposed = DecomposedType(type_obj)  # type: ignore[arg-type]
-        if decomposed.is_optional:
-            return decomposed.get_optional_type().type
-        return type_obj
+
+        def compute():
+            decomposed = DecomposedType(type_obj)  # type: ignore[arg-type]
+            if decomposed.is_optional:
+                return decomposed.get_optional_type().type
+            return type_obj
+
+        return self._get_or_set_cache("_extract_non_optional_type_cache", compute, key=type_obj)

--- a/packages/nvidia_nat_core/src/nat/utils/type_utils.py
+++ b/packages/nvidia_nat_core/src/nat/utils/type_utils.py
@@ -21,7 +21,6 @@ import os
 import sys
 import types
 import typing
-from functools import lru_cache
 from typing import TypeAlias
 
 from pydantic import BaseModel
@@ -57,6 +56,30 @@ else:
         return func
 
 
+class read_only_cached_property:
+
+    def __init__(self, func):
+        self.func = func
+        self.attrname = func.__name__
+        self.__doc__ = func.__doc__
+
+    def __get__(self, instance, owner=None):
+        if instance is None:
+            return self
+        if self.attrname in instance.__dict__:
+            return instance.__dict__[self.attrname]
+        value = self.func(instance)
+        instance.__dict__[self.attrname] = value
+        return value
+
+    def __set__(self, instance, value):
+        raise AttributeError(f"'{self.attrname}' is read-only")
+
+    def __delete__(self, instance):
+        if self.attrname in instance.__dict__:
+            del instance.__dict__[self.attrname]
+
+
 class DecomposedType:
 
     def __init__(self, original: type):
@@ -66,8 +89,7 @@ class DecomposedType:
 
         self.type = original
 
-    @property
-    @lru_cache
+    @read_only_cached_property
     def origin(self):
         """
         Get the origin of the current type using `typing.get_origin`. For example, if the current type is `list[int]`,
@@ -81,8 +103,7 @@ class DecomposedType:
 
         return typing.get_origin(self.type)
 
-    @property
-    @lru_cache
+    @read_only_cached_property
     def args(self):
         """
         Get the arguments of the current type using `typing.get_args`. For example, if the current type is `list[int,
@@ -96,8 +117,7 @@ class DecomposedType:
 
         return typing.get_args(self.type)
 
-    @property
-    @lru_cache
+    @read_only_cached_property
     def root(self):
         """
         Get the root type of the current type. This is the type without any annotations or async generators.
@@ -110,8 +130,7 @@ class DecomposedType:
 
         return self.origin if self.origin is not None else self.type
 
-    @property
-    @lru_cache
+    @read_only_cached_property
     def is_empty(self):
         """
         Check if the current type is eqivalent to `NoneType`.
@@ -123,8 +142,7 @@ class DecomposedType:
         """
         return self.type is types.NoneType
 
-    @property
-    @lru_cache
+    @read_only_cached_property
     def is_class(self):
         """
         Check if the current type is a class using `inspect.isclass`. For example, `list[int]` would return False, but
@@ -138,8 +156,7 @@ class DecomposedType:
 
         return inspect.isclass(self.type)
 
-    @property
-    @lru_cache
+    @read_only_cached_property
     def is_generic(self):
         """
         Check if the current type is a generic using `typing.GenericMeta`. For example, `list[int]` would return True,
@@ -153,8 +170,7 @@ class DecomposedType:
 
         return self.origin is not None
 
-    @property
-    @lru_cache
+    @read_only_cached_property
     def is_annotated(self):
         """
         Check if the current type is an annotated type using `typing.Annotated`. For example, `Annotated[int, str]`
@@ -168,8 +184,7 @@ class DecomposedType:
 
         return self.origin is typing.Annotated
 
-    @property
-    @lru_cache
+    @read_only_cached_property
     def is_union(self):
         """
         Check if the current type is a union type using `typing.Union`. For example, `Union[int, str]` would return
@@ -183,8 +198,7 @@ class DecomposedType:
 
         return self.origin in (typing.Union, types.UnionType)
 
-    @property
-    @lru_cache
+    @read_only_cached_property
     def is_async_generator(self):
         """
         Check if the current type is an async generator type. For example, `AsyncGenerator[int]` would return True,
@@ -202,8 +216,7 @@ class DecomposedType:
             types.AsyncGeneratorType,
         )
 
-    @property
-    @lru_cache
+    @read_only_cached_property
     def is_optional(self):
         """
         Check if the current type is an optional type. For example, `Optional[int]` and `int | None` would return True,
@@ -217,8 +230,7 @@ class DecomposedType:
 
         return self.is_union and types.NoneType in self.args
 
-    @property
-    @lru_cache
+    @read_only_cached_property
     def has_base_type(self):
         """
         Check if the current type has a base type, ignoring any annotations or async generators.

--- a/packages/nvidia_nat_core/tests/nat/observability/mixin/test_type_introspection_mixin.py
+++ b/packages/nvidia_nat_core/tests/nat/observability/mixin/test_type_introspection_mixin.py
@@ -234,7 +234,7 @@ class TestTypeIntrospectionMixin:
             _ = instance.input_type
 
     def test_properties_cached(self):
-        """Test that properties are cached using lru_cache"""
+        """Test that properties are cached via instance dictionary"""
         instance = ConcreteDirectClass()
 
         # Access properties multiple times
@@ -257,8 +257,6 @@ class TestTypeIntrospectionMixin:
 
         # Mock to remove __orig_bases__
         with patch.object(instance.__class__, '__orig_bases__', []):
-            # Clear cache to force re-evaluation
-            instance._extract_input_output_types.cache_clear()
             with pytest.raises(ValueError):
                 _ = instance.input_type
 

